core-schema/src/spec.rs

```diff
@@ -0,0 +1,100 @@
+//! Spec url handling
+//!
+//! `Spec`s are parsed from URL strings and extract the spec's:
+//!   - **identity**, which is the URL excluding the version specifier,
+//!   - **name**, which is the second-to-last path segment of the URL,
+//!     (typically the name of the bare directive exported by the spec), and
+//!   - **version**, specified in the last URL path segment.
+//!
+//! # Example:
+//! ```
+//! use core_schema::*;
+//! assert_eq!(
+//!   Spec::parse("https://spec.example.com/specA/v1.0")?,
+//!   Spec::new("https://spec.example.com/specA", "specA", (1, 0))
+//! );
+//! Ok::<(), SpecParseError>(())
+//! ```
+use std::borrow::Cow;
+
+use thiserror::Error;
+use url::Url;
+
+use crate::{version, version::Version};
+
+/// Specs contain the `identity`, `name`, and `version` extracted from
+/// a spec URL.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Spec {
+    pub identity: Cow<'static, str>,
+    pub name: Cow<'static, str>,
```

Comment:
I worry it might be a little too easy to get confused between `Spec.name` and `Feature.name` and accidentally use the wrong one, writing some code that works as long as you don't use `as`. How important is it that there be a stand-alone `name` field on `spec`? It looks like the only time we ever read it is when we produce a `Feature` so I think we could just validate that the name is valid at `Spec::parse` time (and save `name` or its start index into a private field?) but make actually getting the name later be a method or something?